e2e/test/iothub/messaging/MessageReceiveE2ETests.cs

```diff
@@ -980,5 +1004,38 @@ private async Task DoNotReceiveMessagesSentBeforeSubscriptionAsync(TestDeviceTyp
             deviceClient.Dispose();
             testDeviceCallbackHandler.Dispose();
         }
+
+        // This test ensures that the SDK does not have this bug again
+        // https://github.com/Azure/azure-iot-sdk-csharp/issues/2218
+        private async Task UnsubscribeDoesNotCauseConnectionStatusEventAsync(TestDeviceType type, Client.TransportType transportType)
+        {
+            using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(Logger, s_devicePrefix, type).ConfigureAwait(false);
+            using DeviceClient deviceClient = testDevice.CreateDeviceClient(transportType);
+            bool lostConnection = false;
+            deviceClient.SetConnectionStatusChangesHandler((status, statusChangeReason) =>
+            {
+                if (status == ConnectionStatus.Disconnected || status == ConnectionStatus.Disconnected_Retrying)
+                {
+                    lostConnection = true;
+                }
+            });
+
+            await deviceClient.OpenAsync().ConfigureAwait(false);
+
+            using var testDeviceCallbackHandler = new TestDeviceCallbackHandler(deviceClient, testDevice, Logger);
+
+            // Subscribe to receive C2D messages over the callback.
+            await testDeviceCallbackHandler.SetMessageReceiveCallbackHandlerAsync().ConfigureAwait(false);
+
+            // This will make the client unsubscribe from the mqtt c2d topic/close the amqp c2d link. Neither event
+            // should close the connection as a whole, though.
+            await deviceClient.SetReceiveMessageHandlerAsync(null, null);
+
+            await Task.Delay(1000);
```

Comment:
The disconnection that would happen prior to this fix would not necessarily cause a connection status callback event within the ```deviceClient.SetReceiveMessageHandlerAsync(null, null)``` call, so this task.delay accounts for that.

Reply:
ConfigureAwait(false)